### PR TITLE
Introduce Cypress

### PR DESCRIPTION
**Pull request by joe-bell**
> **WORK IN PROGRESS**

## Description

A working branch to integrate Cypress e2e testing:
1. Installs Cypress with the default template, along with extra tweaks to work with both PNPM and TypeScript

### Checklist

- [ ] The documentation reflects the changes.
- [ ] I have manually tested the application to make sure the changes don’t cause any downstream issues, which includes making sure `./orchest status --ext` is not reporting failures when Orchest is running.
- [ ] In case I changed code in the `orchest-sdk`, I updated its version according to [SemVer](https://semver.org/) in its `_version.py` and updated the version compability table in its `README.md`

- [ ] In case I changed one of the services’ `models.py` I have performed the appropriate database migrations.

### CODE REVIEW

File: .github/workflows/integration-tests.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      - name: Pre cleanup
+        run: |
+          ./orchest stop
+          [ ! -z "$(docker ps -a -q)" ] && docker rm $(docker ps -a -q)
+          sudo bash scripts/clean_userdir.sh
+
+      - name: Build and Start Orchest
+        run: |
+          bash scripts/build_container.sh
+          ./orchest start
```
Comment (yannickperrenet) [inline .github/workflows/integration-tests.yml:54]: Might be good to keep in mind if we want to test the `--cloud` configuration as well.

Comment (fruttasecca) [reply .github/workflows/integration-tests.yml:54]: Indeed, we can either use `--cloud` by default or scale up the number of jobs. Let's see what the run times looks like for 1 browser 1 start mode.

File: .github/workflows/integration-tests.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      # In case some failed job has left state.
+      - name: Pre cleanup
+        run: |
+          ./orchest stop
+          [ ! -z "$(docker ps -a -q)" ] && docker rm $(docker ps -a -q)
+          sudo bash scripts/clean_userdir.sh
+
+      - name: Build and Start Orchest
+        run: |
+          bash scripts/build_container.sh
```
Comment (yannickperrenet) [inline .github/workflows/integration-tests.yml:53]: I think this is indeed the way to go, but we have to keep in mind the fact that Docker could incorrectly build due to cache issues. Although this has not happened in a while now.

Comment (fruttasecca) [reply .github/workflows/integration-tests.yml:53]: Yeah, that issue is indeed a pain to deal with. Not sure we can avoid it without having ephemeral instances.

File: .github/workflows/integration-tests.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+          CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
+          # Recommended: pass the GitHub token lets this action correctly
+          # determine the unique run id necessary to re-run the checks
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Post cleanup
+        run: |
+          ./orchest stop
+          [ ! -z "$(docker ps -a -q)" ] && docker rm $(docker ps -a -q)
+          sudo bash scripts/clean_userdir.sh
```
Comment (yannickperrenet) [inline .github/workflows/integration-tests.yml:73]: Regarding cleanup, who is responsible for cleaning up testcase state? And what if the tests exit unexpectedly?

Could become a problem where e.g. an environment image sticks around and tests failing. Or because the images stick around, the docker installation blows up leaving us no disk space on the instance.

Comment (fruttasecca) [reply .github/workflows/integration-tests.yml:73]: Right now the only state that this action is not taking care of (I think) are environment builds and services (because a service might be a pulled image). The former could be taken care of using tags, the latter requires a bit more smartness. Good catch.

**Comment by joe-bell** [comment]
⚠️ **THIS PR IS NOT READY FOR REVIEW**

This is going to be a pretty big PR, so to avoid bloating the commit trail let's leave reviews until the PR leaves its "DRAFT" status

cc @yannickperrenet @fruttasecca

**Comment by fruttasecca** [comment]
Update:

A few issues have been found & fixed during the development of this PR:
- a race condition that led to dangling environment images in the case of a very quick abortion of the build
- a race condition that led to dangling environment images if the abortion of a build arrived too late to the worker
- a race condition occurring when multiple project deletions for the same project were issued
- a category of issues related to rejected asynchronous client requests has been found, which led to client errors in case of view changes. The issue itself is not very disrupting, but it made integration testing more difficult. Instances of this issue have been fixed by `catching` the promise/
- the path validation for moving a pipeline has been improved
- the path validation for creating a step has been improved, and it is now consistent with the action of pipeline moving by the fact that it also takes care of creating directories that do not exist. Moreover, the path of a step does not being with a `/` anymore, which is consistent with pipeline paths.

**Comment by yannickperrenet** [comment]
I will do a more in-depth review later, but before I forget I wanted to note that we should improve the docs on how to run the tests locally. Which in turn can also help us to maintain the runner that we have.

**Comment by fruttasecca** [comment]
I'll update the docs with a section related to the integration tests. (or maybe tests in general)

File: cypress/fixtures/custom/pipelines/data-passing/B.ipynb [hunk 4]
```diff
@@ -0,0 +1,57 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
```
Comment (yannickperrenet) [inline cypress/fixtures/custom/pipelines/data-passing/B.ipynb:4]: Wouldn't it be easier to use Python scripts here instead of Notebooks? Or are we particularly testing for Notebooks? Keen to hear your thoughts.

File: cypress/integration/examples/actions.spec.ts [hunk 5]
```diff
@@ -0,0 +1,313 @@
+context("Actions", () => {
+  beforeEach(() => {
+    cy.visit("https://example.cypress.io/commands/actions");
+  });
```
Comment (yannickperrenet) [inline cypress/integration/examples/actions.spec.ts:4]: When it comes the `examples` directory, can't we delete it?

File: docs/source/developer_guide/testing.rst [hunk 6]
```diff
@@ -0,0 +1,98 @@
+.. _testing:
+
+Testing
+=======
```
Comment (yannickperrenet) [inline docs/source/developer_guide/testing.rst:4]: As mentioned in https://github.com/orchest/orchest/pull/384#issuecomment-903853423 the docs will be revised. Just mentioning here so that we can have a dedicated thread.

Comment (yannickperrenet) [reply docs/source/developer_guide/testing.rst:4]: I think we should add something to the "development workflow" docs that either links to these "testing" section or notes how they can be run.

File: scripts/run_integration_tests.sh [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+			[y]* ) break;;
+			[n]* ) exit;;
+			* ) echo "${msg}";;
+		esac
+	done
+fi
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+cd "${DIR}/../"
+
```
Comment (yannickperrenet) [inline scripts/run_integration_tests.sh:56]: > In my local Docker environment I still had a `orchest-env...` image which made all the `beforeEach` fail

Should we also add something for this in this script? Would definitely make running the script a lot easier without having to first check the docs.

Comment (fruttasecca) [reply scripts/run_integration_tests.sh:56]: Yes, the script could actually take care of deleting all environments so that this is taken care of and doesn't need to be referenced in the docs.

File: cypress/integration/examples/actions.spec.ts (hunk 5 above)
Comment (fruttasecca) [reply cypress/integration/examples/actions.spec.ts:4]: Yes, we can indeed delete it. We initially went for not deleting it, happy to delete it in hindsight.

File: scripts/run_integration_tests.sh (hunk 7 above)
Comment (fruttasecca) [reply scripts/run_integration_tests.sh:56]: Moreover, the script could actually check if orchest is running and reject if it's not.

Comment (yannickperrenet) [reply scripts/run_integration_tests.sh:56]: > Moreover, the script could actually check if orchest is running and reject if it's not.

I like this addition. What if the script starts Orchest in case it is not running yet? Similarly it shouldn't reject if there is still an `orchest-env` instead it should just clean it up and run the tests.

File: cypress/integration/examples/actions.spec.ts (hunk 5 above)
Comment (yannickperrenet) [reply cypress/integration/examples/actions.spec.ts:4]: I don't see any reason to leave in all this code that isn't even ours.

File: cypress/fixtures/custom/pipelines/data-passing/B.ipynb (hunk 4 above)
Comment (yannickperrenet) [reply cypress/fixtures/custom/pipelines/data-passing/B.ipynb:4]: Let's start with Notebooks and if we end up changing them often we can revisit the decision and move to scripts :+1:

File: cypress/integration/file-system-interactions.spec.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import { PROJECTS_DIR, SAMPLE_PROJECT_NAMES } from "../support/common";
+
+describe("file system interactions", () => {
```
Comment (yannickperrenet) [inline cypress/integration/file-system-interactions.spec.ts:3]: How about adding:
* Deleting a pipeline through the FS

@ricklamers This would also be a question for you, but why do we strictly enforce a naming convention on the project names if they are created through the UI but still "detect" FS projects that don't follow the convention:
1. Try to create a project named `hello world` through the UI --> ERROR
2. Try to create a project named `hello world` through the FS --> UI picks up and calls `git init` inside the detected project

File: cypress/integration/interactive-runs.spec.ts [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+  STEPS,
+  PIPELINES,
+  SAMPLE_PROJECT_NAMES,
+  SAMPLE_PIPELINE_NAMES,
+  SAMPLE_STEP_NAMES,
+  setStepParameters,
+} from "../support/common";
+
+describe("interactive runs", () => {
+  beforeEach(() => {
```
Comment (yannickperrenet) [inline cypress/integration/interactive-runs.spec.ts:14]: (Not saying we have to add all these tests in this PR, just so we can write them down somewhere.)

I think we are missing some tests around the interaction between interactive runs and environments. For example, what if a used environment is currently building. A bit more of the edge cases.

File: cypress/integration/pipelines.spec.ts [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+    // pipeline.
+    let originalPath = `${PROJECTS_DIR}/${SAMPLE_PROJECT_NAMES.P1}/${expectedFile}`;
+    let copyPath = `${PROJECTS_DIR}/${SAMPLE_PROJECT_NAMES.P1}/copy.orchest`;
+    cy.exec(`cp ${originalPath} ${copyPath}`);
+    // Reload to force the discovery.
+    cy.visit("/pipelines");
+    cy.findAllByTestId(TEST_ID.PIPELINES_TABLE_ROW).should("have.length", 2);
+  });
+
+  it("deletes a pipeline through the fs", () => {
```
Comment (yannickperrenet) [inline cypress/integration/pipelines.spec.ts:72]: I see the pipeline FS interaction is here. Maybe we should change the names of the modules as I don't think it is clear that this test is here instead of in the other. ([reference](https://github.com/orchest/orchest/pull/384#discussion_r695489820))

File: cypress/integration/file-system-interactions.spec.ts (hunk 8 above)
Comment (fruttasecca) [reply cypress/integration/file-system-interactions.spec.ts:3]: >Deleting a pipeline through the FS

This is already being tested but I ended up placing it in the wrong file (`pipelines.spec.ts`), will move it.

>Moving a project through the FS (essentially just renaming it)

Will add

>why do we strictly enforce a naming convention on the project names if they are created through the UI but still "detect" FS projects that don't follow the convention

This is a good point,  we should make an issue out of it, as this PR is very large already.

Comment (yannickperrenet) [reply cypress/integration/file-system-interactions.spec.ts:3]: The issue: https://github.com/orchest/orchest/issues/441

File: cypress/integration/interactive-runs.spec.ts (hunk 9 above)
Comment (fruttasecca) [reply cypress/integration/interactive-runs.spec.ts:14]: I have added this case to our internal documentation.

File: docs/source/developer_guide/testing.rst (hunk 6 above)
Comment (fruttasecca) [reply docs/source/developer_guide/testing.rst:4]: I have updated the testing docs, given the update to the `run_integration_tests.sh` script I haven't found a lot to add, let me know if anything comes up to you.

**Review by yannickperrenet — APPROVED** [review]

**Comment by yannickperrenet** [comment]
LGTM!
